src/runtime/virtcontainers/clh.go

```diff
@@ -178,8 +178,11 @@ var clhKernelParams = []Param{
 }
 
 var clhDebugKernelParams = []Param{
-	{"console", "ttyS0,115200n8"},     // enable serial console
-	{"systemd.log_target", "console"}, // send loggng to the console
+	// Both of these console options are required!
+	{"console", "ttyS0,115200n8"}, // for kernel boot console messages
```

Comment:
So you're saying this is for kernel boot messages, but later you always set `Serial` to `Null`. That means you're gonna loose all these logs.
If the point is to get both early kernel messages along with the agent logs redirected to the console, then I don't understand why the initial code wasn't working.
I mean you want a kernel command line with `console=ttyS0`, along with the device being properly emulated.
So unless I'm missing something, it seems to me like the tty might not be handled/redirected correctly from Kata itself, but I don't think we need to change anything in this code here.

Reply:
@sboeuf - fwics, `ttyS0` captures all guest kernel boot messages with CLH. But you _only_ get boot messages: no userland output is displayed. For that, you seemingly must specify `hvc0`. But you need both specified on the kernel cli to ensure we get all output in the Kata logs. You can see (or not see) what is happening using something like this:

```bash
$ sudo curl --unix-socket /tmp/clh.sock -i \
     -X PUT 'http://localhost/api/v1/vm.create'  \
     -H 'Accept: application/json'               \
     -H 'Content-Type: application/json'         \
     -d '{
         "cpus":{"boot_vcpus": 4, "max_vcpus": 4},
         "kernel":{"path":"/opt/kata/share/kata-containers/vmlinux.container"},
         "cmdline":{"args":"ro agent.log=debug console=ttyS0 root=/dev/vda1 rootflags=data=ordered,errors=remount-ro noreplace-smp no_timer_check panic=1 rootfstype=ext4 systemd.log_target=console systemd.unit=kata-containers.target"},
         "disks":[{"path":"/opt/kata/share/kata-containers/kata-containers.img"}],
         "rng":{"src":"/dev/urandom"},
         "net":[{"ip":"192.168.10.10", "mask":"255.255.255.0", "mac":"12:34:56:78:90:01"}],
           "serial": {
    "iommu": true,
    "mode": "Tty"
  }, "console": {
    "iommu": false,
    "mode": "Off"
  }}'
```

Aside from all the CLH debug, you'll see the kernel boot messages ending with `Run /sbin/init as init process`, but you will not see any userland output.

However, try with this:

```bash
$   sudo curl --unix-socket /tmp/clh.sock -i \
     -X PUT 'http://localhost/api/v1/vm.create'  \
     -H 'Accept: application/json'               \
     -H 'Content-Type: application/json'         \
     -d '{
         "cpus":{"boot_vcpus": 4, "max_vcpus": 4},
         "kernel":{"path":"/opt/kata/share/kata-containers/vmlinux.container"},
         "cmdline":{"args":"ro agent.log=debug console=ttyS0 console=hvc0 root=/dev/vda1 rootflags=data=ordered,errors=remount-ro noreplace-smp no_timer_check panic=1 rootfstype=ext4 systemd.log_target=console systemd.unit=kata-containers.target"},
         "disks":[{"path":"/opt/kata/share/kata-containers/kata-containers.img"}],
         "rng":{"src":"/dev/urandom"},
         "net":[{"ip":"192.168.10.10", "mask":"255.255.255.0", "mac":"12:34:56:78:90:01"}],
           "serial": {
    "iommu": false,
    "mode": "Null"
  }, "console": {
    "iommu": false,
    "mode": "Tty"
  }}'
```

There you will also see the final `Run /sbin/init as init process` guest kernel boot message. But you'll also see systemd and job output:

```
[    0.347503] Run /sbin/init as init process
systemd 219 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 -SECCOMP +BLKID +ELFUTILS +KMOD +IDN)
Detected virtualization kvm.
Detected architecture x86-64.

Welcome to ^[[0;31mCentOS Linux 7 (Core)^[[0m!

    :

[^[[32m  OK  ^[[0m] Listening on D-Bus System Message Bus Socket.
[^[[32m  OK  ^[[0m] Reached target Sockets.
[^[[32m  OK  ^[[0m] Reached target Basic System.
[^[[32m  OK  ^[[0m] Started Kata Containers Agent.
         Starting NTP client/server...
{"msg":"announce","level":"INFO","ts":"2021-09-29T09:04:08.630361372+00:00","subsystem":"root","name":"kata-agent","source":"agent","pid":"79","version":"0.1.0","agent-version":"2.3.0-alpha0","api-version":"0.0.1","agent-type":"rust","agent-commit":"2.3.0-alpha0-4d46d71e157002a55c4371e5ad348fbec8c36ed1","config":"AgentConfig { debug_console: false, dev_mode: false, log_level: Debug, hotplug_timeout: 3s, debug_console_vport: 0, log_vport: 0, container_pipe_size: 0, server_addr: \"vsock://-1:1024\", unified_cgroup_hierarchy: false, tracing: Disabled }"}
{"msg":"starting uevents handler","level":"INFO","ts":"2021-09-29T09:04:08.631899123+00:00","version":"0.1.0","subsystem":"uevent","source":"agent","name":"kata-agent","pid":"79"}
{"msg":"ttRPC server started","level":"INFO","ts":"2021-09-29T09:04:08.632110172+00:00","source":"agent","pid":"79","version":"0.1.0","name":"kata-agent","subsystem":"rpc","address":"vsock://-1:1024"}
[^[[32m  OK  ^[[0m] Started NTP client/server.
[^[[32m  OK  ^[[0m] Reached target Kata Containers Agent Target.
```

But unless this is a bug in CLH, you need both devices as `hvc0` doesn't provide kernel output and `ttyS0` doesn't provide userland output.

Reply:
I don't understand, I thought `systemd.log_target=console` was responsible for sending the agent logs to the available console(s), and in case ttyS0 is available I don't see why the logs don't land there :-/
Also, i was looking at `fc.go` and realized the debug case uses only `console=ttyS0`, how do you get the agent logs in this case?
Still about `fc.go`, I've also noticed that the non-debug case was adding `agent.log_vport`, which we don't use in CH, why is that?
There are lots of differences across QEMU/FC/CH about the way they gather logs, which is confusing and makes it hard to understand why it works for QEMU and FC compared to CH.

Reply:
This is looking like a CLH bug - I booted a VM using CLH standalone, specifying `init=/usr/bin/fake-init.sh`. That script echoes to `stdout`, `stderr`, `/dev/tty`, `/dev/console` and `$(tty)`. If I boot specifying `console=ttyS0`, I get no kernel or userland output. But if I boot using `console=ttyS0 console=hvc0`, I see kernel output, and all output from my fake init program. But if I boot using just `console=hvc0`, I actually also get all kernel + userland output (including all output from my fake init script) - aka no data lost fwics, atleast for this trivial test example.

ftr, if I enable the console and the serial device in the CLH config, I get a `VmBoot(VmBoot(ConfigValidation(DoubleTtyMode)))` error.

Reply:
Regarding fc, others will have to comment (@mcastelino or @devimc maybe?) I agree we should try to standardise and refactor console handling, where possible, for all the hypervisors.

Reply:
@jcvenegas - it's worth noting that Kata boots QEMU using `console=hvc0 console=hvc1` currently.

I've dug into this further and I can get full output with cloud-hypervisor v18.0 if I configure the serial device (without using `hvc0`), but **only** when redirected to a file:

| kernel CLI | CLH console | CLH serial | kernel output | userland output |
|-|-|-|-|-|
| `console=ttyS0` | `Null` | `Tty` | yes | no |
| `console=ttyS0` | `Null` | `File` | yes | **yes** |


@sboeuf, @rbradford - can you confirm these findings? If so, I'll raise a CLH bug.

Reply:
I've found the issue! Kata boots CLH with `-vv` which asks it to log at debug level. When that happens, after the kernel hands off to the init process, there is some sort of irq storm resulting in CLH logging messages like the following furiously:

```
[    0.396839] Run /usr/bin/fake-init-test-console.sh as init process
1cloud-hypervisor: 9.200179928s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
:cloud-hypervisor: 9.200217886s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
 cloud-hypervisor: 9.200244985s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
/cloud-hypervisor: 9.200282134s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ucloud-hypervisor: 9.200307563s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
scloud-hypervisor: 9.200334872s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
rcloud-hypervisor: 9.200363124s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
/cloud-hypervisor: 9.200388067s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
bcloud-hypervisor: 9.200412957s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
icloud-hypervisor: 9.200438587s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ncloud-hypervisor: 9.200469438s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
/cloud-hypervisor: 9.200497919s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
fcloud-hypervisor: 9.200523749s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
acloud-hypervisor: 9.20054806s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
kcloud-hypervisor: 9.200575609s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ecloud-hypervisor: 9.200602683s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
-cloud-hypervisor: 9.200648382s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
icloud-hypervisor: 9.200673285s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ncloud-hypervisor: 9.200697721s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
icloud-hypervisor: 9.200722643s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
tcloud-hypervisor: 9.200748108s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
-cloud-hypervisor: 9.200774682s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
tcloud-hypervisor: 9.200804232s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ecloud-hypervisor: 9.200831758s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
scloud-hypervisor: 9.200858851s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
tcloud-hypervisor: 9.200885249s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
-cloud-hypervisor: 9.200911613s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ccloud-hypervisor: 9.200936676s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ocloud-hypervisor: 9.200963678s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ncloud-hypervisor: 9.200986055s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
scloud-hypervisor: 9.201010172s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ocloud-hypervisor: 9.201033204s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
lcloud-hypervisor: 9.201072244s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
ecloud-hypervisor: 9.201099269s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
.cloud-hypervisor: 9.201125158s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
scloud-hypervisor: 9.201156576s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
hcloud-hypervisor: 9.201188409s: <vcpu2> DEBUG:devices/src/ioapic.rs:394 -- Interrupt successfully delivered
```

Now, look at the first byte of each line. Taken together they make:

```
1: /usr/bin/fake-init-test-console.sh
```

So the problem is that CLH itself is corrupting the serial output with its own debug messages. Disabling debug output by running CLH either without `-v` or with only a single `-v` (info log level) resolves the problem! So I think the ultimate solution is to run the CLH binary with `-vv --log-file /some/where/clh-output.log` as that way we can get sane serial output and a separate log of CLH messages. Does that sound reasonable @sboeuf, @rbradford, @jcvenegas?

Reply:
Or you could send the serial to a pty or file. Also I strongly recommend not running with "-vv" the level of output is not helpful.

Reply:
Also, if you can use the virtio-console (hvc0) over serial then you will have a much faster boot too. (Use `--console <wherever> --serial off` for best performance but only do that if you do not have `console=ttyS0` in the kernel command line.)

Reply:
> Also I strongly recommend not running with "-vv" the level of output is not helpful.

Agreed. It seems to be overkill for our needs. Which makes the minimal fix for this PR to get all output (guest kernel + userland):

```diff
diff --git a/src/runtime/virtcontainers/clh.go b/src/runtime/virtcontainers/clh.go
index 170f76f4..4dc19207 100644
--- a/src/runtime/virtcontainers/clh.go
+++ b/src/runtime/virtcontainers/clh.go
@@ -936,7 +936,7 @@ func (clh *cloudHypervisor) launchClh() (int, error) {
                // and in a nested environment this could increase
                // the chances to fail because agent is not
                // ready on time.
-               args = append(args, "-vv")
+               args = append(args, "-v")
        }
```

1 byte change ftw! :smile: